pkg/scheduler/objects/application.go

```diff
@@ -604,6 +630,121 @@ func (sa *Application) tryAllocate(headRoom *resources.Resource, nodeIterator fu
+		}
+		// walk over the placeholders, allow for processing all as we can have multiple task groups
+		phAllocs := sa.getPlaceholderAllocations()
+		for _, ph := range phAllocs {
+			// we could have already released this placeholder and are waiting for the shim to confirm
+			// and check that we have the correct task group before trying to swap
+			if ph.released || request.taskGroupName != ph.taskGroupName {
+				continue
+			}
+			if phFit == nil && reqFit == nil {
+				phFit = ph
+				reqFit = request
+			}
+			// this is inefficient but needs to do for now as we only have the ID not the handle on the node itself
+			var ok bool
+			var node *Node
+			for iterator.HasNext() {
+				node, ok = iterator.Next().(*Node)
+				if !ok {
+					log.Logger().Warn("Node iterator failed to return a node")
+					return nil
+				}
+				if node.NodeID == ph.NodeID {
+					break
+				}
+			}
+			// got the node run same checks as for reservation (all but fits)
+			// resource usage should not change anyway between placeholder and real one
+			if node != nil && node.preReserveConditions(request.AllocationKey) {
+				alloc := NewAllocation(common.GetNewUUID(), node.NodeID, request)
+				// double link to make it easier to find
+				alloc.Releases = []*Allocation{ph}
+				ph.Releases = []*Allocation{alloc}
+				alloc.Result = Replaced
+				// mark placeholder as released
+				ph.released = true
+				_, err := sa.updateAskRepeatInternal(request, -1)
+				if err != nil {
+					log.Logger().Warn("ask repeat update failed unexpectedly",
+						zap.Error(err))
+				}
+				return alloc
+			}
+			iterator.Reset()
+		}
+	}
+	// we checked all placeholders and asks nothing worked as yet
+	// pick the first fit and try all nodes if that fails give up
+	if phFit != nil && reqFit != nil {
+		for iterator.HasNext() {
+			node, ok := iterator.Next().(*Node)
+			if !ok {
+				log.Logger().Warn("Node iterator failed to return a node")
+				return nil
+			}
+			if err := node.preAllocateCheck(reqFit.AllocatedResource, reservationKey(nil, sa, reqFit), false); err != nil {
+				continue
+			}
+			// skip the node if conditions can not be satisfied
+			if !node.preAllocateConditions(reqFit.AllocationKey) {
+				continue
+			}
+			// allocation worked: on a non placeholder node update result and return
+			alloc := NewAllocation(common.GetNewUUID(), node.NodeID, reqFit)
+			// double link to make it easier to find
+			alloc.Releases = []*Allocation{phFit}
+			phFit.Releases = []*Allocation{alloc}
+			alloc.Result = Replaced
+			// mark placeholder as released
+			phFit.released = true
+			// update just the node to make sure we keep its spot
+			// no queue update as we're releasing the placeholder and are just temp over the size
+			if !node.AddAllocation(alloc) {
+				log.Logger().Debug("Node update failed unexpectedly",
+					zap.String("applicationID", sa.ApplicationID),
+					zap.String("ask", reqFit.String()),
+					zap.String("placeholder", phFit.String()))
+				return nil
```

Comment:
We do not need to wait for a placeholder to release before updating the node resources?
This seems diverse from the normal processing logic. 
If we can merge this part to the normal processing code, can we do something like the following:

1. Placeholder allocate fails (doesn't fit)
2. Mark a placeholder to release, "park" the request
3. Wait until the placeholder is released in the shim, do normal allocation for the request

This means: the placeholder wasn't good, so we try to schedule the pod as before.

Reply:
> We do not need to wait for a placeholder to release before updating the node resources?
> This seems diverse from the normal processing logic.
> If we can merge this part to the normal processing code, can we do something like the following:
> 
> 1. Placeholder allocate fails (doesn't fit)
> 2. Mark a placeholder to release, "park" the request
> 3. Wait until the placeholder is released in the shim, do normal allocation for the request
> 
> This means: the placeholder wasn't good, so we try to schedule the pod as before.

Just add to this. One concern I have is we may go over queue resource quota with this.
Imagine a case we have 10 placeholders and 10 real pods for an app, at the beginning we have 10 placeholders allocated, where we have 

10 placeholders 1GB * 10 = 10GB
  - app: 10GB in placeholders
  - nodes: 10GB in allocated resources
  - queue: 10GB allocated resources

imagine the worst-case scenario, the placeholders don't fit, and we start to fallback allocating the real pods on other nodes. Then we get

  - app: 10GB in placeholders, 10GB in allocated resources
  - nodes: 20GB
  - queue: 10GB (* dangerous)

*dangerous: both app and nodes are tracking 20GB used resources for this app, but for the queue, it is 10GB. If the queue has a resource quota set to 20GB. It might be possible that before shim releases all 10 placeholders, the core will schedule other pods from other apps, that will goes over the quota defined for this queue.

Reply:
That would break the gang. If the placeholder is allocated the queue usage is reserved for the gang. When we release the placeholder there is no guarantee that another application in the queue is not going to take that space. This could be an older application or even a newer application.
It would in (almost) all cases directly break gang scheduling

Reply:
hi @wilfred-s even with the approach in this PR, gang scheduling is not 100% guaranteed. If we can't utilize the placeholders for the real pods and schedule them on some other nodes. We do not know how many of them can be allocated successfully and how many will keep pending. It still breaks the gang scheduling semantics.

The case I described above is pretty bad. The scheduler data should always be consistent, few dimensions are: used resources by all apps in a queue should equal to used resources of this queue. This is a more serious problem than breaking gang scheduling in the worst-case scenario.

Reply:
Unless you can in some way guarantee that a node cannot change over time there is no way that you can get that 100% guarantee for gang scheduling. This is just not possible, we're close to it but not there. What this approach tries to do is making sure that the exception case is a small as possible and handled as well as possible. Preventing breakage where possible.

An early release of a placeholder from the queue's total use will break it at the point that we have told the application that the gang is reserved. In the case that there is a limit on the queue it could even mean that we cannot get that quota back as an other application in the queue has started using it. It might thus cause such a problem that the gang can never be fully allocated.
The only other option is to reserve the node so we do not lose the node. Then on the release confirmation from the placeholder finish the real allocation. The release would be processed on the node and in the app. We would need to push the real allocation through a special allocation cycle to update the "new" node. Then update the app with the final result.
We would also somehow make sure that the reservation is removed from the normal `TryReserveAllocate` cycle adding another complexity. The node that got reserved will be unavailable for scheduling all the time the release is being processed. That also does not feel correct and really complex.

For the temporary discrepancy: it *only* happens if all the placeholder run on nodes that can not for some reason support the real allocation (it is the failure case). It lasts only for the time it takes K8s to release the placeholder and it is all back in sync again.